doc/read-the-docs-site/adr/0006-indexer-resuming-strategy.rst

```diff
@@ -45,16 +45,89 @@ Decision
+  We will provide as intersection points the points returned from the following pseudo-code:
+
+  .. code-block:: haskell
+
+    -- | Query the resumable points of each indexer then select the resumable
+    -- points of the indexer that has the lowest immutable point. If two indexer
+    -- have the same lowest immutable point, select the one which has the lowest
+    -- overall point (immutable or not).
+    selectIntersectionPoints indexers = do
+      resumingPoints <- mapM queryResumingPoints indexers
+      head $ sortOn (\rps -> (maximum $ filter isImmutablePoint rps, maximum rps)) resumingPoints
+
+  Next, we need a function which processes each ``ChainSyncEvent`` returned by the chain-sync client for each indexer.
+  The pseudo-code is as follows:
+
+  .. code-block:: haskell
+
+    -- Preconditions:
+    --   * The 'initialLastSyncPoints' param should be sorted in ascending order on SlotNo
+    --   * Each element of the 'initialLastSyncPoints' param should satisfy the property 'getSlotNo point == getSlotNo (point + 1) + 1'.
+    applyChainSyncEvent initialLastSyncPoints previousChainSyncEventPoint event indexer = do
+      case event of
+        Resume point -> do
+          maybePoint <- find (\p -> getSlotNo p == getSlotNo point) initialLastSyncPoints
+          case maybePoint of
+            Nothing ->
+              pure () -- Do nothing
+            Just _ ->
+              update initialLastSyncPoints $ filter (\p -> getSlotNo p > getSlotNo point)
+        RollBackward point -> do
+          rewind point indexer
+        RollForward block point -> do
+          case initialLastSyncPoints of
+            -- Processed all initial last sync points. The indexer is in sync.
+            -- So we just index the block.
+            [] -> do
+              insert block indexer
+
+            -- The lowest last sync point is the same as the current chain event point.
+            -- We drain the event and remove the last sync point from the initial last sync points.
+            (lowestLastSyncPoint:restOfLastSyncPoints) | point == lowestLastSyncPoint -> do
+                update initialLastSyncPoints $ \_ -> restOfLastSyncPoints
+
+            -- The lowest last sync point is not the same as the current chain event point, but they
+            -- have the same slot number.
+            -- We rewind to the previous point, insert the block and remove all points from the
+            -- initial last sync point list.
+            -- We drain the event and remove the last sync point from the initial last sync points.
+            (lowestLastSyncPoint:_) | getSlotNo point == getSlotNo lowestLastSyncPoint -> do
+                rewind previousChainSyncEventPoint indexer
+                insert block indexer
+                update initialLastSyncPoints $ \_ -> []
+
+            -- The chain sync event slot number is higher than the lowest last sync point of the
+            -- indexer. Should not happen though.
```

Comment:
Do all the blocks contain the same amount of slots I thought so but IIRC. it doesn't look alike in cardano-scan.

Reply:
If they are different though, it probably means that a we need to rollback to the previous block, so I would suggest a rollback to the previous block too in this case.

Reply:
By "Do all the blocks contain the same amount of slots", do you mean if each slot produces a block? Then the answer is no.

Also, are you talking about this line specifically?
```
rewind previousChainSyncEventPoint indexer
```

Reply:
No I meant "Do each block contain the same amount of slots" if it not the case and if a rollback to block `N` happen, the new block `N+1` may have a different size than the previous block `N+1` and thus the slot number wont match.

Reply:
And I was proposing to add a `rollback` before the insert in the lines below this comment (L110).

Reply:
I still don't understand what you mean. Blocks don't contain slots. Blocks are created a specific slot.

Reply:
Sorry, I wasn't specific enough: I was talking about the distance (in slots) between two blocks.

Reply:
Ok, but the distance between slots seems irrelevant to me in this situation.

The only situation where we get in this case statement is if we started the chain-sync client with a resuming point that is higher than this indexer's resuming point. If that happens, then there's a bug in the implementation. 

For example, if we get a `RollForward` event in point `10`, but the lowest resuming point of the indexer is `5`, then the initial points we provided to the chain-sync client are wrong. I guess I should probably just do an `error "..."` here?